Cosmos DB: Operation Level Metrics (#1438)
open-telemetry/semantic-conventions: docs/database/cosmosdb.md

## Operation Level Metrics

The following metrics provide insights into Azure Cosmos DB client operation performance and behavior.

### Metric: `db.client.operation.duration`

This metric is [required][MetricRequired].

It captures the total time taken by an Azure Cosmos DB operation. This metric follows the common [db.client.operation.duration](/docs/database/database-metrics.md#metric-dbclientoperationduration) definition.

Refer [db.cosmosdb.operation.request_charge](#metric-dbcosmosdboperationrequest_charge) metrics for dimensions.

### Metric: `db.client.response.row_count`

This metric is [required][MetricRequired].

It captures the number of items returned by a query or feed operation in Azure Cosmos DB. It helps identify response sizes that may contribute to high latency, increased memory/CPU usage, or network call failures. This metric follows the common [db.client.response.row_count](/docs/database/database-metrics.md#metric-dbclientresponserow_count) definition.

Refer [db.cosmosdb.operation.request_charge](#metric-dbcosmosdboperationrequest_charge) metrics for dimensions.

### Metric: `db.cosmosdb.operation.request_charge`

This metric is [required][MetricRequired].

> It captures the Request Units consumed by each operation in Azure Cosmos DB. Since Request Units serve as a form of currency within the Azure Cosmos DB database, monitoring their usage is crucial to avoid throttling.

Reviewer: ```suggestion
It captures the Request Units consumed by each operation in Azure Cosmos DB. Since Request Units serve as a form of throughput control within the Azure Cosmos DB database, monitoring their usage is crucial to avoid throttling.
```

| Value  | Description | Stability |
|---|---|---|
| `BoundedStaleness` | bounded_staleness | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `ConsistentPrefix` | consistent_prefix | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `Eventual` | eventual | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `Session` | session | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `Strong` | strong | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

`error.type` has the following list of well-known values. If one of them applies, then the respective value MUST be used; otherwise, a custom value MAY be used.

| Value  | Description | Stability |
|---|---|---|
| `_OTHER` | A fallback error value to be used when the instrumentation doesn't define a custom value. | ![Stable](https://img.shields.io/badge/-stable-lightgreen) |

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

### Metric: `db.cosmosdb.client.active_instances`

This metric is [required][MetricRequired].

> It captures the number of active instances at any given time. Best practices dictate that there should ideally be only one instance of the SDK client per process. Having multiple instances of the SDK client can lead to CPU or memory-related issues.

Reviewer: There are some multi-tenant scenarios where multiple clients may be required. Let's specify one client per account instead of one per process

```suggestion
It captures the number of active instances at any given time. Best practices dictate that there should ideally be only one instance of the SDK client per Azure Cosmos DB account. Having multiple instances of the SDK client for the same account in a single process can lead to CPU or memory-related issues.
```

| `db.operation.name`                  | `"read_item"` |
| `server.address`                     | `"account.documents.azure.com"` |
| `db.cosmosdb.client_id`              | `"3ba4827d-4422-483f-b59f-85b74211c11d"` |
| `user_agent.original`                | `"cosmos-netstandard-sdk/3.23.0\|3.23.1\|1\|X64\|Linux 5.4.0-1098-azure 104 18\|.NET Core 3.1.32\|S\|"` |
| `db.cosmosdb.request_content_length` | `20` |
| `db.response.status_code`            | `201` |
| `db.cosmosdb.sub_status_code`        | `0` |
| `db.cosmosdb.request_charge`         | `7.43` |

## Operation Level Metrics

> The following metric instruments describe Azure Cosmos DB client Operation level behaviour.

Reviewer 2: nit:

maybe

```suggestion
The following metrics provide insights into Azure Cosmos DB client operation performance and behavior.
```

?

<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

### Metric: `db.cosmosdb.client.active_instances`

This metric is [required][MetricRequired].

It captures the number of active instances at any given time. Best practices dictate that there should ideally be only one instance of the SDK client per process. Having multiple instances of the SDK client can lead to CPU or memory-related issues.

<!-- semconv metric.db.cosmosdb.client.active_instances -->
<!-- NOTE: THIS TEXT IS AUTOGENERATED. DO NOT EDIT BY HAND. -->
<!-- see templates/registry/markdown/snippet.md.j2 -->
<!-- prettier-ignore-start -->
<!-- markdownlint-capture -->
<!-- markdownlint-disable -->

| Name     | Instrument Type | Unit (UCUM) | Description    | Stability |
| -------- | --------------- | ----------- | -------------- | --------- |
| `db.cosmosdb.client.active_instances` | UpDownCounter | `{instance}` | Number of active client instances | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|

> | [`db.collection.name`](/docs/attributes-registry/db.md) | string | Cosmos DB container name. [1] | `public.users`; `customers` | `Conditionally Required` If available | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: is collection name associated with a client? I think you don't provide a container name when  you create a client. Or would this metric be incremented every time someone does `db.GetContainer()` ?

Author: For every cosmos db call, we always have the container and database information. Hence, we are using it as a dimension.

| [`db.cosmosdb.consistency_level`](/docs/attributes-registry/db.md) | string | Account or request [consistency level](https://learn.microsoft.com/azure/cosmos-db/consistency-levels). | `Eventual`; `ConsistentPrefix`; `BoundedStaleness`; `Strong`; `Session` | `Conditionally Required` If available. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | [`db.cosmosdb.sub_status_code`](/docs/attributes-registry/db.md) | int | Cosmos DB sub status code. | `1000`; `1002` | `Conditionally Required` when response was received and contained sub-code. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: this attribute does not look right on this metric - single client will receive multiple responses.

Author: What do you mean by saying a single client will receive multiple responses? Each operation will return a single response. We aim to evaluate the performance of these operations based on different parameters such as consistency level, operation name, substatus code, etc. Therefore, we are using these as metric dimensions.

> | [`db.namespace`](/docs/attributes-registry/db.md) | string | The name of the database, fully qualified within the server address and port. | `customers`; `test.users` | `Conditionally Required` If available. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: same concerns as on the `db.collection.name` - is it available on the client level or would you expect the metric to be incremented each time `client.GetDatabase` is called?

Author: Yes, it is always available that which container and database, we are querying

Author: Now instance count metrics has, only account information as dimension

> | [`db.operation.name`](/docs/attributes-registry/db.md) | string | The name of the operation or command being executed. [2] | `findAndModify`; `HMSET`; `SELECT` | `Conditionally Required` [3] | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: same concern - there is no single operation that can be put on this metric

Author: removed...now instance size count only account information as dimension

> | [`db.response.status_code`](/docs/attributes-registry/db.md) | string | Database response status code. [4] | `102`; `ORA-17002`; `08P01`; `404` | `Conditionally Required` [5] | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: and here

Author: removed...now instance size count only account information as dimension

> | [`error.type`](/docs/attributes-registry/error.md) | string | Describes a class of error the operation ended with. [6] | `timeout`; `java.net.UnknownHostException`; `server_certificate_invalid`; `500` | `Conditionally Required` If and only if the operation failed. | ![Stable](https://img.shields.io/badge/-stable-lightgreen) |

Reviewer 2: and here

Author: removed...now instance size has only account information as dimension
